### PR TITLE
Add check for 0-dimensional array arguments in hvncat and produce an error

### LINKED ISSUE #41047: `hvncat` miscomputes output bounds for empty Vectors
On: Version 1.7.0-DEV.1222 (2021-06-01)
Commit 4a8572fc6d

Tried to figure out the new syntax for hvncat and ran into this bug:
```julia
julia> [[];;7;8]
2×2 Matrix{Any}:
 1  #undef
 2  #undef
```

Not sure what the expected behavior is, but since `[7;8;;[]]` throws an error, I assume that some boundscheck is missing.

This seems to be an almost equivalent call
```julia
 julia> hvncat(((1,2),(3,)), false, zeros(Int,0,), 7, 8)
2×2 Matrix{Int64}:
 7  0
 8  0
```

Trying to debug, these works as expected:
```julia
julia> hvncat(((1,2),(3,)), false, zeros(Int,2,0), 7.0, 8.0)
2×1 Matrix{Float64}:
 7.0
 8.0
```
```julia
julia> hvncat(((1,2),(3,)), false, zeros(Int, 2,), 7, 8)
2×2 Matrix{Int64}:
 0  7
 0  8
```

However, you can quickly cause a segfault:
```julia
julia> hvncat(((1,2),(3,)), false, zeros(Int,0,0,0), 7, 8)
```
<details>
  <summary>Trace:</summary>

```julia
julia> hvncat(((1,2),(3,)), false, zeros(Int,0,0,0), 7, 8)
signal (11): Segmentation fault
in expression starting at none:0
jl_gc_pool_alloc at /work/mattiasf/juliagit_dev/src/gc.c:1211
jl_gc_alloc_ at /work/mattiasf/juliagit_dev/src/julia_internal.h:315 [inlined]
jl_gc_alloc at /work/mattiasf/juliagit_dev/src/gc.c:3278
_new_array_ at /work/mattiasf/juliagit_dev/src/array.c:122 [inlined]
_new_array at /work/mattiasf/juliagit_dev/src/array.c:188 [inlined]
jl_alloc_array_1d at /work/mattiasf/juliagit_dev/src/array.c:459
Array at ./boot.jl:453 [inlined]
Array at ./boot.jl:472 [inlined]
getindex at ./array.jl:411 [inlined]
with_repl_linfo at /work/mattiasf/juliagit_dev/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:503
_jl_invoke at /work/mattiasf/juliagit_dev/src/gf.c:2244 [inlined]
jl_apply_generic at /work/mattiasf/juliagit_dev/src/gf.c:2426
display at /work/mattiasf/juliagit_dev/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:254
display at /work/mattiasf/juliagit_dev/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:266
_jl_invoke at /work/mattiasf/juliagit_dev/src/gf.c:2244 [inlined]
jl_apply_generic at /work/mattiasf/juliagit_dev/src/gf.c:2426
display at ./multimedia.jl:328
_jl_invoke at /work/mattiasf/juliagit_dev/src/gf.c:2244 [inlined]
jl_apply_generic at /work/mattiasf/juliagit_dev/src/gf.c:2426
jl_apply at /work/mattiasf/juliagit_dev/src/julia.h:1760 [inlined]
jl_f__call_latest at /work/mattiasf/juliagit_dev/src/builtins.c:751
#invokelatest#2 at ./essentials.jl:726 [inlined]
invokelatest at ./essentials.jl:724 [inlined]
print_response at /work/mattiasf/juliagit_dev/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:288
#45 at /work/mattiasf/juliagit_dev/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:272
jfptr_YY.45_20661 at /work/mattiasf/juliagit_dev/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /work/mattiasf/juliagit_dev/src/gf.c:2244 [inlined]
jl_apply_generic at /work/mattiasf/juliagit_dev/src/gf.c:2426
with_repl_linfo at /work/mattiasf/juliagit_dev/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:505
_jl_invoke at /work/mattiasf/juliagit_dev/src/gf.c:2244 [inlined]
jl_apply_generic at /work/mattiasf/juliagit_dev/src/gf.c:2426
print_response at /work/mattiasf/juliagit_dev/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:270
do_respond at /work/mattiasf/juliagit_dev/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:841
jfptr_do_respond_20618 at /work/mattiasf/juliagit_dev/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /work/mattiasf/juliagit_dev/src/gf.c:2244 [inlined]
jl_apply_generic at /work/mattiasf/juliagit_dev/src/gf.c:2426
jl_apply at /work/mattiasf/juliagit_dev/src/julia.h:1760 [inlined]
jl_f__call_latest at /work/mattiasf/juliagit_dev/src/builtins.c:751
#invokelatest#2 at ./essentials.jl:726 [inlined]
invokelatest at ./essentials.jl:724 [inlined]
run_interface at /work/mattiasf/juliagit_dev/usr/share/julia/stdlib/v1.7/REPL/src/LineEdit.jl:2493
jfptr_run_interface_20024 at /work/mattiasf/juliagit_dev/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /work/mattiasf/juliagit_dev/src/gf.c:2244 [inlined]
jl_apply_generic at /work/mattiasf/juliagit_dev/src/gf.c:2426
run_frontend at /work/mattiasf/juliagit_dev/usr/share/julia/stdlib/v1.7/REPL/src/REPL.jl:1227
#49 at ./task.jl:411
jfptr_YY.49_20803 at /work/mattiasf/juliagit_dev/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /work/mattiasf/juliagit_dev/src/gf.c:2244 [inlined]
jl_apply_generic at /work/mattiasf/juliagit_dev/src/gf.c:2426
jl_apply at /work/mattiasf/juliagit_dev/src/julia.h:1760 [inlined]
start_task at /work/mattiasf/juliagit_dev/src/task.c:822
Allocations: 10796169 (Pool: 10791934; Big: 4235); GC: 13
[1]    3143063 segmentation fault (core dumped)  ./julia
```

</details>

Edit: Bug generalizes to higher dims: `[zeros(Int,2,0);;;7;8]`

### PR DESCRIPTION
**Pull request by BioTurboNick**
Resolves #41047 by checking for zero-length inputs, accounting for the case where the shape form of `hvncat` has higher-dimensional arguments than the highest dimension of concatenation, and appropriate tests for all.

### CODE REVIEW

File: base/abstractarray.jl [hunk 1]
```diff
@@ -2196,8 +2196,11 @@ function _typed_hvncat(::Type{T}, ::Val{N}, as...) where {T, N}
     # into the destination
     nd = N
     Ndim = 0
-    for a ∈ as
+    for i ∈ eachindex(as)
+        a = as[i]
```
Comment (simeonschaub) [inline base/abstractarray.jl:2200]: Why this change? Doesn't seem like `i` is used anywhere else.

Comment (BioTurboNick) [reply base/abstractarray.jl:2200]: For the error message to specify the index. Guess I could have done `enumerate()`

Comment (simeonschaub) [reply base/abstractarray.jl:2200]: Ah, I didn't see that. In that case it's fine.

File: base/abstractarray.jl [hunk 2]
```diff
@@ -2299,8 +2302,9 @@ function _typed_hvncat(::Type{T}, shape::Tuple{Vararg{Tuple, N}}, row_first::Boo
     shapepos = ones(Int, nd)
 
     for i ∈ eachindex(as)
+        length(as[i]) > 0 || ArgumentError("argument $i has no elements") |> throw
```
Comment (mbauman) [inline base/abstractarray.jl:2305]: This is the smallest of nits, but since you use the functional form above let's at least be consistent within one patch :)

```suggestion
        length(as[i]) > 0 || throw(ArgumentError("argument $i has no elements"))
```

File: base/abstractarray.jl [hunk 3]
```diff
@@ -2313,7 +2317,7 @@ function _typed_hvncat(::Type{T}, shape::Tuple{Vararg{Tuple, N}}, row_first::Boo
 
             wasstartblock = blockcounts[d] == 1 # remember for next dimension
 
-            isendblock = blockcounts[d] == shape[d][shapepos[d]]
+            isendblock = blockcounts[d] == shape[d > N ? end : d][shapepos[d]]
```
Comment (mbauman) [inline base/abstractarray.jl:2320]: And while I'm here, I find this slightly better since I had to double-check that `N == length(shape)`.

```suggestion
            isendblock = blockcounts[d] == shape[min(end, d)][shapepos[d]]
```

**Review by mbauman — COMMENTED** [review]
This is outside the commentable diff, but:

https://github.com/JuliaLang/julia/blob/de1444c02be3d225eb13190719457017d038c534/base/abstractarray.jl#L2295

Is it possible for `as` to be empty?

File: base/abstractarray.jl (hunk 2 above)
Comment (BioTurboNick) [reply base/abstractarray.jl:2305]: In the other PR I shifted everything else to `throw(...)`, so thanks :-)

**Review by mbauman — APPROVED** [review]

**Comment by DilumAluthge** [comment]
@mbauman CI is green now. Is this good to merge?

File: base/abstractarray.jl [hunk 4]
```diff
@@ -2299,8 +2302,9 @@ function _typed_hvncat(::Type{T}, shape::Tuple{Vararg{Tuple, N}}, row_first::Boo
     shapepos = ones(Int, nd)
 
     for i ∈ eachindex(as)
+        length(as[i]) > 0 || throw(ArgumentError("argument $i has no elements"))
```
Comment (JeffBezanson) [inline base/abstractarray.jl:2305]: I don't see why this is a special case, i.e. why isn't it covered by the other check that the number of elements matches?

Comment (JeffBezanson) [reply base/abstractarray.jl:2305]: For example `[zeros(1,2,0);;;7 8]` doesn't need to be an error.

Comment (BioTurboNick) [reply base/abstractarray.jl:2305]: The former check is the length of as; this is the length of each element. Though I should probably be using cat_length here in case any of them are strings.

But you're right. This is a crude check. The dimension calculating algorithm right now just can't handle arrays with a zero dimension. e.g. [zeros(1, 0);;;7 8] creates an underfilled array. Seems better to make them error for now until I can address that.

**Comment by BioTurboNick** [comment]
I have a better solution, which will end up in #41143, that allows correct zero-length arrays to work the way you would expect. So this PR can be closed.
